File: ios/Classes/MuxDataPlugin.swift
import Flutter
import UIKit

public class MuxDataPlugin: NSObject, FlutterPlugin {
    private var muxStats: MuxStats?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "plugins.flutter.io/mux_data", binaryMessenger: registrar.messenger())
        let instance = MuxDataPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]

        if (call.method == "init") {
            guard muxStats == nil else {
                result(FlutterError(code: "", message: "MuxStats handler already initialized.", details: nil))
                return
            }
            muxStats = MuxStats(environmentKey: args["environmentKey"] as! String, viewerId: args["viewerId"] as! String)
            result(nil)
            return
        }

        guard let muxStats = muxStats else {
            result(FlutterError(code: "", message: "Uninitialized MuxStats handler.", details: nil))
            return
        }

        print(call.method)

        switch (call.method) {
        case "playerReady":
            muxStats.playerReady()
            break
        case "viewInit":
            muxStats.viewInit()
            break
        case "videoChange":
            break
        case "play":
            muxStats.play()
            break
        case "playing":
            muxStats.playing()
            break
        case "paused":
            muxStats.pause()
            break
        case "timeUpdate":
            muxStats.timeUpdate(args!["playheadTimeMillis"] as! Int)
            break
        case "seeking":
            muxStats.seeking()
            break
        case "seeked":
            muxStats.seeked()
            break
        case "rebufferStart":
            muxStats.rebufferStart()
            break
        case "rebufferEnd":
            muxStats.rebufferEnd()
            break
        case "error":
            muxStats.error(code: args!["code"] as! Int, message: args!["message"] as! String)
            break
        case "ended":
            muxStats.ended()
            break
        case "renditionChange":
            break
        case "orientationChange":
            break
        case "viewEnd":
            break
        default:
            result(FlutterMethodNotImplemented)
            return
        }

        result(nil)
    }
}

File: android/src/main/java/com/example/mux_data/MuxDataPlugin.java
package com.example.mux_data;

import android.content.Context;
import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** MuxDataPlugin */
public class MuxDataPlugin implements FlutterPlugin, MethodCallHandler {
    private MethodChannel channel;
    private Context context;
    private PlayerListener listener;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "plugins.flutter.io/mux_data");
        channel.setMethodCallHandler(this);
        context = flutterPluginBinding.getApplicationContext();
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
        listener = null;
        channel = null;
        context = null;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("init")) {
            listener = new PlayerListener(context, call.argument("environmentKey"), call.argument("viewerId"));
            result.success(null);
            return;
        }
        if (listener == null) {
            result.error("", "Uninitialized MuxStats handler.", null);
            return;
        }

        switch (call.method) {
            case "play":
                listener.play();
                break;
            case "playing":
                listener.playing();
                break;
            case "pause":
                listener.pause();
                break;
            case "timeUpdate":
                listener.timeUpdate(call.argument("playheadTimeMillis"));
                break;
            case "seeking":
                listener.seeking();
                break;
            case "seeked":
                listener.seeked();
                break;
            case "rebufferStart":
                listener.rebufferStart();
                break;
            case "rebufferEnd":
                listener.rebufferEnd();
                break;
            case "error":
                listener.error(call.argument("code"), call.argument("message"));
                break;
            case "ended":
                listener.ended();
                break;
            default:
                result.notImplemented();
                return;
        }

        result.success(null);
    }
}
